### PR TITLE
Add Enabled flag for attributes to metadata.yaml

### PR DESCRIPTION
**Pull request by hughesjj**
**Description:** Adds an Enabled configuration option to metric generators so that receiver authors can add new attributes without forcing all existing consumers to use such. 

**Link to tracking Issue:** [16373](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/16373)

**Testing:** `make mdatagen-test && make generate`
To verify I hit all of the `resource_attributes` in the various `metadata.yaml`, I installed `yq` and ran the below

```bash
rm resource_attr_metadata_yaml.log && grep -R --files-with-matches "resource_attributes" **/metadata.yaml | xargs yq .resource_attributes - >> resource_attr_metadata_yaml.log && echo "$(grep description resource_attr_metadata_yaml.log | wc -l) vs $(grep enabled resource_attr_metadata_yaml.log | wc -l)"
97 vs 97
```
I find it best to run something like this for new commits to this PR
`make mdatagen-test && make generate && make otelcontribcol && git add -u && git commit -m "Meaningful update text" && git push`

**Documentation:** None, other than changes to generated documentation.

### CODE REVIEW

File: cmd/mdatagen/loader.go [hunk 1]
```diff
@@ -158,12 +158,22 @@ type attribute struct {
 	Description string `validate:"notblank"`
 	// NameOverride can be used to override the attribute name.
 	NameOverride string `mapstructure:"name_override"`
+	// Enabled defines whether the attribute is enabled by default.
+	Enabled *bool `yaml:"enabled" validate:"required"`
```
Comment (bogdandrutu) [inline cmd/mdatagen/loader.go:162]: Why pointer?

Comment (dmitryax) [reply cmd/mdatagen/loader.go:162]: This was actually needed during the transition from the old metrics builder. Now it can be a bool value, not a pointer. I'll submit a PR for metric.Enabled

Comment (dmitryax) [reply cmd/mdatagen/loader.go:162]: I removed the pointer from metrics in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/16775. @hughesjj you probably will need to rebase

File: cmd/mdatagen/metric-metadata.yaml [hunk 2]
```diff
@@ -22,6 +24,8 @@ attributes:
     name_override:
     # Required: description of the attribute.
     description:
+    # Optional: Whether the attribute is emitted by default.
+    enabled: <bool>
```
Comment (dmitryax) [inline cmd/mdatagen/metric-metadata.yaml:28]: Let's not add it to the metric attributes yet. That will be done as part of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/16374

File: cmd/mdatagen/testdata/documentation.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+| resource.attr.explicitly.disabled | Explicitly disabled ResourceAttribute. | Str  | false |
+| string.resource.attr.withvalue | ResourceAttribute that has hardcoded value set. | Str | true |
+| bool.resource.attr | ResourceAttribute with a boolean value. | Bool | true |
+
+## Metric attributes
+
+| Name | Description | Values |
+| --- | --- | --- |
+| cpu_type (type) | The type of CPU consumption | user, io_wait, system |
+| host | The type of CPU consumption |  |
```
Comment (dmitryax) [inline cmd/mdatagen/testdata/documentation.md:37]: This shouldn't be manually added. It's generated by make `mdatagen-test` now

File: cmd/mdatagen/main_test.go [hunk 4]
```diff
@@ -35,6 +35,35 @@ func Test_runContents(t *testing.T) {
...
+    description: ResourceAttribute that has hardcoded value set.
+    enabled: true
+    value: state
+    type: string
+
+  bool.resource.attr:
+    description: ResourceAttribute with a boolean value.
+    enabled: true
+    value: "0"
+    type: bool
```
Comment (dmitryax) [inline cmd/mdatagen/main_test.go:65]: Let's not add it here. This is just a smoke test with the simplest valid yaml

File: cmd/mdatagen/loader.go [hunk 5]
```diff
@@ -182,12 +182,35 @@ type attribute struct {
...
+	Enabled bool `yaml:"enabled" validate:"required"`
+	// Value can optionally specify the value this resource attribute will have.
+	// For example, the resource attribute may have the identifier `MemState` to its
+	// value may be `state` when used.
+	Value string
+	// Enum can optionally describe the set of values to which the resource attribute can belong.
+	Enum []string
+	// Type is a resource attribute type.
+	Type ValueType `mapstructure:"type"`
+}
```
Comment (dmitryax) [inline cmd/mdatagen/loader.go:212]: Why do we need another structure?  We can reuse `attribute` I believe

File: cmd/mdatagen/metric-metadata.yaml [hunk 6]
```diff
@@ -7,9 +7,14 @@ sem_conv_version: 1.9.0
 
 # Optional: map of resource attribute definitions with the key being the attribute name.
 resource_attributes:
-  <attribute.name>:
+  <resource_attribute.name>:
+    # Optional: this field can be used to override the actual attribute name defined by the key.
+    # It should be used if multiple metrics have different attributes with the same name.
+    name_override:
```
Comment (dmitryax) [inline cmd/mdatagen/metric-metadata.yaml:13]: `name_override` is not needed for resource attributes

File: cmd/mdatagen/metadata.yaml [hunk 7]
```diff
@@ -7,13 +7,32 @@ sem_conv_version: 1.9.0
...
 
+  resource.attr.explicitly.disabled:
+    description: Explicitly disabled ResourceAttribute
+    enabled: false
+    type: string
+
+  string.resource.attr.withvalue:
+    description: ResourceAttribute that has hardcoded value set.
+    enabled: true
+    value: state
```
Comment (dmitryax) [inline cmd/mdatagen/metadata.yaml:27]: What is the `value`? I don't believe it's needed

File: cmd/mdatagen/metadata.yaml [hunk 8]
```diff
@@ -7,13 +7,32 @@ sem_conv_version: 1.9.0
...
   string.enum.resource.attr:
     description: Resource attribute with a known set of string values.
+    enabled: true
     type: string
     enum: [one, two]
 
+  resource.attr.explicitly.disabled:
+    description: Explicitly disabled ResourceAttribute
+    enabled: false
+    type: string
```
Comment (dmitryax) [inline cmd/mdatagen/metadata.yaml:22]: Let's keep just this one as an addittional attribute. Should be anough

File: cmd/mdatagen/templates/metrics.go.tmpl [hunk 9]
```diff
@@ -93,6 +134,38 @@ var MapAttribute{{ $name.Render }} = map[string]Attribute{{ $name.Render }}{
...
+
+		// MapResourceAttribute{{ $name.Render }} is a helper map of string to ResourceAttribute{{ $name.Render }} resource attribute value.
+		var MapResourceAttribute{{ $name.Render }} = map[string]ResourceAttribute{{ $name.Render }}{
+		{{- range $info.Enum }}
+			"{{ . }}": ResourceAttribute{{ $name.Render }}{{ . | publicVar }},
+		{{- end }}
+		}
+
+	{{ end }}
+{{- end }}
```
Comment (dmitryax) [inline cmd/mdatagen/templates/metrics.go.tmpl:168]: I don't think we need this API. We just need to update `With<resource attribute>` options to read the settings and don't add the attribute if it's disabled

File: Makefile [hunk 10]
```diff
@@ -247,6 +247,7 @@ generate:
 mdatagen-test:
 	cd cmd/mdatagen && $(GOCMD) install .
 	cd cmd/mdatagen && $(GOCMD) generate ./...
+	cd cmd/mdatagen && $(GOCMD) test ./...
```
Comment (dmitryax) [inline Makefile:250]: Maybe better put it in the first place? Feel free to submit a separate PR to unblock the automated builds :)

File: cmd/mdatagen/loader_test.go [hunk 11]
```diff
@@ -36,17 +36,26 @@ func Test_loadMetadata(t *testing.T) {
...
 					},
 					"string.enum.resource.attr": {
 						Description: "Resource attribute with a known set of string values.",
+						Enabled:     true,
 						Enum:        []string{"one", "two"},
 						Type: ValueType{
 							ValueType: pcommon.ValueTypeStr,
 						},
 					},
+					"resource.attr.explicitly.disabled": {
```
Comment (dmitryax) [inline cmd/mdatagen/loader_test.go:52]: Let's stick with the same default/optional terms as for metrics
```suggestion
					"optional.resource.attr": {
```

File: cmd/mdatagen/main.go [hunk 12]
```diff
@@ -78,6 +78,9 @@ func generateFile(tmplFile string, outputFile string, md metadata) error {
 				"publicVar": func(s string) (string, error) {
 					return formatIdentifier(s, true)
 				},
+				"resourceAttributeInfo": func(an attributeName) attribute {
+					return md.ResourceAttributes[an]
+				},
```
Comment (dmitryax) [inline cmd/mdatagen/main.go:83]: I don't think it's needed, and you don't use it

File: cmd/mdatagen/loader_test.go [hunk 13]
```diff
@@ -36,17 +36,26 @@ func Test_loadMetadata(t *testing.T) {
...
+						Description: "Explicitly disabled ResourceAttribute",
```
Comment (dmitryax) [inline cmd/mdatagen/loader_test.go:53]:
```suggestion
						Description: "Resource attribute disabled by default",
```

File: cmd/mdatagen/templates/metrics.go.tmpl [hunk 14]
```diff
@@ -60,6 +60,51 @@ func DefaultMetricsSettings() MetricsSettings {
...
+	ras.enabledProvidedByUser = parser.IsSet("enabled")
+	return nil
+}
+
+// ResourceAttributesSettings provides settings for {{ .Name }} resource attributes.
+type ResourceAttributesSettings struct {
+	{{- range $name, $attr := .ResourceAttributes }}
+	{{ $name.Render }} ResourceAttributeSettings `mapstructure:"{{ $name }}"`
+	{{- end }}
+}
```
Comment (dmitryax) [inline cmd/mdatagen/templates/metrics.go.tmpl:96]: I think we should wrap `MetricsSettings` and `ResourceAttributesSettings` into another structure like  `MetricsBuilderConfig` and add it to the component configs with `mapstructure:",squash"`. And it should be passed to the `NewMetricsBuilder` instead of `MetricsSettings` and `WithResourceAttributesSettings`.

File: cmd/mdatagen/metadata.yaml [hunk 15]
```diff
@@ -8,30 +8,41 @@ resource_attributes:
...
+  optional.resource.attr:
+    description: explicitly disabled ResourceAttribute
+    type: string
+    enabled: false
 
 attributes:
   string_attr:
     description: Attribute with any string value.
     type: string
+    enabled: true
```
Comment (dmitryax) [inline cmd/mdatagen/metadata.yaml:28]: please don't add this to metric attributes

File: cmd/mdatagen/templates/documentation.md.tmpl [hunk 16]
```diff
@@ -83,11 +83,19 @@ metrics:
...
 ## Resource Attributes
 
+| Name | Description | Type | Enabled |
+| ---- | ----------- | ---- | ------- |
+{{- range $attributeName, $attributeInfo := .ResourceAttributes }}
+| {{ $attributeName }} | {{ $attributeInfo.Description }} | {{ $attributeInfo.Type }} | {{ $attributeInfo.Enabled }} |
+{{- end }}
+{{- end }}
+
+## Metric attributes
```
Comment (dmitryax) [inline cmd/mdatagen/templates/documentation.md.tmpl:93]: Please remove this section. Metrics attributes are rendered separately for each metric

File: cmd/mdatagen/templates/metrics.go.tmpl [hunk 17]
```diff
@@ -55,6 +55,47 @@ func DefaultMetricsSettings() MetricsSettings {
...
 
+// ResourceAttributeSettings provides common settings for a particular metric.
+type ResourceAttributeSettings struct {
+	Enabled bool `mapstructure:"enabled"`
+
+	enabledProvidedByUser bool
+}
+
+// IsEnabledProvidedByUser returns true if `enabled` option is explicitly set in user settings to any value.
+func (ras *ResourceAttributeSettings) IsEnabledProvidedByUser() bool {
```
Comment (dmitryax) [inline cmd/mdatagen/templates/metrics.go.tmpl:66]: This method isn't needed anymore. It was intended to be exposed for scraper authors to show warnings based on user configuration but the warnings are now built into the builder

File: cmd/mdatagen/loader.go [hunk 18]
```diff
@@ -186,6 +186,12 @@ type attribute struct {
 	Description string `mapstructure:"description"`
 	// NameOverride can be used to override the attribute name.
 	NameOverride string `mapstructure:"name_override"`
+	// Enabled defines whether the attribute is enabled by default.
+	Enabled bool `yaml:"enabled" validate:"required"`
+	// Value can optionally specify the value this attribute will have.
+	// For example, the attribute may have the identifier `MemState` to its
+	// value may be `state` when used.
+	Value string
```
[the remark this answers is not in the document]
Comment (dmitryax) [reply cmd/mdatagen/loader.go:194]: Can you remove this?

File: cmd/mdatagen/testdata/documentation.md [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+| Name | Description       | Type | Enabled |
+| --- | --- | --- | --- |
+| hwid | motherboard hw id | Str  | true    |
+
+## Metric attributes
+
+| Name            | Description                 | Values                |
+| --- | --- | --- |
+| cpu_type (type) | The type of CPU consumption | user, io_wait, system |
+| host            | The type of CPU consumption |                       |
```
Comment (dmitryax) [inline cmd/mdatagen/testdata/documentation.md:33]: Is this not up to date?

File: .chloggen/add_resource_attribute_settings.yaml [hunk 20]
```diff
@@ -0,0 +1,18 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
```
[the remark this answers is not in the document]
Comment (dmitryax) [reply .chloggen/add_resource_attribute_settings.yaml:2]: `op` is an internal function, not a breaking change

File: cmd/mdatagen/loader_test.go [hunk 21]
```diff
@@ -36,17 +36,26 @@ func Test_loadMetadata(t *testing.T) {
...
 					"string.enum.resource.attr": {
 						Description: "Resource attribute with a known set of string values.",
+						Enabled:     true,
 						Enum:        []string{"one", "two"},
 						Type: ValueType{
 							ValueType: pcommon.ValueTypeStr,
 						},
 					},
+					"optional.resource.attr": {
+						Description: "explicitly disabled ResourceAttribute",
```
Comment (dmitryax) [inline cmd/mdatagen/loader_test.go:53]: nit
```suggestion
						Description: "Explicitly disabled ResourceAttribute.",
```

File: cmd/mdatagen/templates/metrics_test.go.tmpl [hunk 22]
```diff
@@ -1,5 +1,5 @@
 // Code generated by mdatagen. DO NOT EDIT.
-
+{{- $sep := "" }}{{- $assignSign := ":=" }}
```
Comment (dmitryax) [inline cmd/mdatagen/templates/metrics_test.go.tmpl:2]: Why is this change needed?

File: cmd/mdatagen/templates/metrics_test.go.tmpl [hunk 23]
```diff
@@ -103,20 +103,27 @@ func TestMetricsBuilder(t *testing.T) {
 
 			assert.Equal(t, 1, metrics.ResourceMetrics().Len())
 			rm := metrics.ResourceMetrics().At(0)
-			{{- $assignSign := ":=" }}
+			{{- $assignSign = ":=" }}
 			attrCount := 0
+			enabledAttrCount := 0
 			{{- range $name, $info := .ResourceAttributes }}
-				attrCount++
+				enabled {{ $assignSign }} mb.resourceAttributesSettings.{{ $name.Render }}.Enabled
```
Comment (dmitryax) [inline cmd/mdatagen/templates/metrics_test.go.tmpl:110]: `enabled` is an ambiguous name. Let's not introduce the variable

File: cmd/mdatagen/templates/metrics_test.go.tmpl [hunk 24]
```diff
@@ -103,20 +103,27 @@ func TestMetricsBuilder(t *testing.T) {
...
 				attrVal, ok {{ $assignSign }} rm.Resource().Attributes().Get("{{ $name }}")
-				assert.True(t, ok)
-				{{- if $info.Enum }}
-					assert.Equal(t, "{{ index $info.Enum 0 }}", attrVal.Str())
-				{{- else }}
-					assert.EqualValues(t, {{ $info.Type.TestValue }}, attrVal.{{ $info.Type }}())
-				{{- end }}
+				attrCount++
+				assert.Equal(t, enabled, ok)
+				if enabled {
```
Comment (dmitryax) [inline cmd/mdatagen/templates/metrics_test.go.tmpl:114]:
```suggestion
				assert.Equal(t, mb.resourceAttributesSettings.{{ $name.Render }}.Enabled, ok)
				if mb.resourceAttributesSettings.{{ $name.Render }}.Enabled {
```

**Review by dmitryax — COMMENTED** [review]
Few nits. Otherwise, LGTM. Please move if from draft once comments are addressed

File: cmd/mdatagen/internal/metadata/generated_metrics_test.go [hunk 25]
```diff
@@ -1,5 +1,4 @@
 // Code generated by mdatagen. DO NOT EDIT.
-
```
Comment (dmitryax) [inline cmd/mdatagen/internal/metadata/generated_metrics_test.go:2]: Sorry for nitpicking. This removed line brings inconsistency with generated_metrics.go. Please remove it, and we can merge.

**Review by dmitryax — APPROVED** [review]
